app/scripts/modules/core/src/presentation/forms/Validation.spec.ts

```diff
@@ -0,0 +1,319 @@
+import {
+  Validation,
+  buildValidators,
+  Validator,
+  ArrayItemValidator,
+  buildValidatorsAsync,
+  IValidationBuilder,
+} from './Validation';
+
+const { isRequired, minValue, maxValue } = Validation;
+
+const makeAsync = (syncValidator: Validator): Validator => {
+  return (value, label) => {
+    const result = syncValidator(value, label);
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        result ? reject(result) : resolve();
+      }, 1);
+    });
+  };
+};
+
+interface ISynchronousTestCase {
+  expectation: string; // feels more like a title but jasmine calls it expectation
+  values: any;
+  expectedResult: any;
+  builders: (builder: IValidationBuilder) => void;
+}
+
+const synchronousTestCases: ISynchronousTestCase[] = [
+  {
+    expectation: 'returns empty errors when validating no validators',
+    values: { foo: 'bar' },
+    builders: builder => {
+      builder.field('foo', 'Foo').validate([]);
+    },
+    expectedResult: {},
+  },
+  {
+    expectation: 'returns correct error when validating top level field',
+    values: {},
+    builders: builder => {
+      builder.field('foo', 'Foo').validate([isRequired()]);
+    },
+    expectedResult: {
+      foo: 'Foo is required.',
+    },
+  },
+  {
+    expectation: 'returns correct error when validating a deep field',
+    values: {},
+    builders: builder => {
+      builder.field('foo.bar.baz', 'Foo').validate([isRequired()]);
+    },
+    expectedResult: {
+      foo: {
+        bar: {
+          baz: 'Foo is required.',
+        },
+      },
+    },
+  },
+  {
+    expectation: 'aggregates multiple levels of errors correctly',
+    values: {},
+    builders: builder => {
+      builder.field('foo', 'Foo').validate([isRequired()]);
+      builder.field('bar.baz', 'Baz').validate([isRequired()]);
+    },
+    expectedResult: {
+      foo: 'Foo is required.',
+      bar: {
+        baz: 'Baz is required.',
+      },
+    },
+  },
+  {
+    expectation: 'validates arrays and aggregates them correctly',
+    values: {
+      lotsastuff: [1, 2, 3, 4, 5],
+    },
+    builders: builder => {
+      const arrayNotEmpty: Validator = (array, label) => array.length < 1 && `${label} must have at least 1 item.`;
+      const { arrayForEach } = builder;
+
+      builder.field('lotsastuff', 'Array').validate([
+        isRequired(),
+        arrayNotEmpty,
```

Comment:
I don't really have a problem with it. I think in practice, `Validator`s will come on many forms:

1. Exported or simple reference (like `arrayNotEmpty`)
2. Returned by a factory for convenience (`minValue(3)` is probably a better example)
3. Inline anonymous function

It would probably look cleanest if they were all references, but that itself is not a good enough reason to have `min3`, `max42` references all over the place.